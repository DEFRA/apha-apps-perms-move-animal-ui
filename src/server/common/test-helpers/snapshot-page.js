import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
