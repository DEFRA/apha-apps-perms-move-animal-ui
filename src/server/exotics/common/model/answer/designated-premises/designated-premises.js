import { RadioButtonAnswer } from '~/src/server/common/model/answer/radio-button/radio-button.js'
/** @import {RadioButtonConfig} from '~/src/server/common/model/answer/radio-button/radio-button.js' */

/**
 * @typedef {{ isDesignatedPremises: string }} DesignatedPremisesPayload
 */

/**
 * @augments {RadioButtonAnswer<DesignatedPremisesPayload>}
 */
export class DesignatedPremisesAnswer extends RadioButtonAnswer {
  /** @type { RadioButtonConfig } */
  static config = {
    payloadKey: 'isDesignatedPremises',
    hint: 'A designated premises has government approval to undertake certain animal-related activities, such as moving eggs.',
    options: {
      yes: { label: 'Yes' },
      no: { label: 'No' },
      unknown: { label: "I don't know" }
    },
    validation: {
      empty: 'Select if the premises is designated'
    }
  }
}
