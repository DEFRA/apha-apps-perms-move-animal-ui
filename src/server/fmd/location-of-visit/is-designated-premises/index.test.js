import { describePageSnapshot } from '~/src/server/common/test-helpers/snapshot-page.js'
import { isDesignatedPremisesPage } from './index.js'
import { DesignatedPremisesAnswer } from '../../common/model/answer/designated-premises/designated-premises.js'
import { AnimalsOnPremisesPage } from '../animals-on-premises/index.js'

const sectionKey = 'locationOfVisit'
const questionKey = 'isDesignatedPremises'
const pageUrl = '/fmd/location-of-visit/designated-premises'
const page = isDesignatedPremisesPage
const question = 'Is the premises designated?'

describe('IsDesignatedPremisesPage', () => {
  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(DesignatedPremisesAnswer)
  })

  describe('nextPage', () => {
    it.each([
      ['yes', AnimalsOnPremisesPage],
      ['no', AnimalsOnPremisesPage],
      ['unknown', AnimalsOnPremisesPage]
    ])('for %s should return %s', (value, expectedPage) => {
      const nextPage = page.nextPage()
      expect(nextPage).toBeInstanceOf(expectedPage)
    })
  })

  describePageSnapshot({
    describes: 'content',
    it: 'should render expected response and content',
    pageUrl
  })
})
