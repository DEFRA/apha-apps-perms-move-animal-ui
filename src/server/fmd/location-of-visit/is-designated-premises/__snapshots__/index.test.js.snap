// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsDesignatedPremisesPage content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset" aria-describedby="isDesignatedPremises-hint">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
      <h1 class="govuk-fieldset__heading">
        Is the premises designated?
      </h1>
    </legend>
    <div id="isDesignatedPremises-hint" class="govuk-hint">
      A designated premises has government approval to undertake certain animal-related activities, such as moving eggs
    </div>
    <div class="govuk-radios" data-module="govuk-radios">
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="isDesignatedPremises" name="isDesignatedPremises" type="radio" value="yes">
        <label class="govuk-label govuk-radios__label" for="isDesignatedPremises">
          Yes
        </label>
      </div>
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="no" name="isDesignatedPremises" type="radio" value="no">
        <label class="govuk-label govuk-radios__label" for="no">
          No
        </label>
      </div>
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="unknown" name="isDesignatedPremises" type="radio" value="unknown">
        <label class="govuk-label govuk-radios__label" for="unknown">
          I don't know
        </label>
      </div>
    </div>
  </fieldset>
</div>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
